File: crates/wadup-guest/src/subcontent.rs
//! File-based sub-content emission for WADUP.
//!
//! Emits sub-content for recursive processing by WADUP using files:
//! - `/subcontent/data_N.bin` - raw data bytes
//! - `/subcontent/metadata_N.json` - metadata (filename, optional offset/length)

use serde::Serialize;
use std::cell::RefCell;
use std::fs::File;
use std::io::Write;

thread_local! {
    static FILE_COUNTER: RefCell<usize> = RefCell::new(0);
}

fn next_counter() -> usize {
    FILE_COUNTER.with(|c| {
        let val = *c.borrow();
        *c.borrow_mut() = val + 1;
        val
    })
}

/// Metadata for sub-content with bytes.
#[derive(Serialize)]
struct SubContentMetadata {
    filename: String,
}

/// Metadata for sub-content slice (references input content).
#[derive(Serialize)]
struct SubContentSliceMetadata {
    filename: String,
    offset: usize,
    length: usize,
}

pub struct SubContent;

impl SubContent {
    /// Emit sub-content bytes for recursive processing.
    ///
    /// Writes data to `/subcontent/data_N.bin` and metadata to `/subcontent/metadata_N.json`.
    /// WADUP processes the sub-content when the metadata file is closed.
    pub fn emit_bytes(data: &[u8], filename: &str) -> Result<(), String> {
        let n = next_counter();
        let data_path = format!("/subcontent/data_{}.bin", n);
        let metadata_path = format!("/subcontent/metadata_{}.json", n);

        // Write data file first
        let mut data_file = File::create(&data_path)
            .map_err(|e| format!("Failed to create subcontent data file '{}': {}", data_path, e))?;
        data_file.write_all(data)
            .map_err(|e| format!("Failed to write subcontent data file '{}': {}", data_path, e))?;
        drop(data_file); // Close data file

        // Write metadata file (triggers processing when closed)
        let metadata = SubContentMetadata {
            filename: filename.to_string(),
        };
        let json = serde_json::to_string(&metadata)
            .map_err(|e| format!("Failed to serialize subcontent metadata: {}", e))?;

        let mut meta_file = File::create(&metadata_path)
            .map_err(|e| format!("Failed to create subcontent metadata file '{}': {}", metadata_path, e))?;
        meta_file.write_all(json.as_bytes())
            .map_err(|e| format!("Failed to write subcontent metadata file '{}': {}", metadata_path, e))?;
        // File closed on drop, triggering WADUP processing

        Ok(())
    }

    /// Emit a slice of the input content as sub-content (zero-copy).
    ///
    /// The slice references a range of the original `/data.bin` content without copying.
    /// Only writes metadata to `/subcontent/metadata_N.json`.
    pub fn emit_slice(offset: usize, length: usize, filename: &str) -> Result<(), String> {
        let n = next_counter();
        let metadata_path = format!("/subcontent/metadata_{}.json", n);

        let metadata = SubContentSliceMetadata {
            filename: filename.to_string(),
            offset,
            length,
        };
        let json = serde_json::to_string(&metadata)
            .map_err(|e| format!("Failed to serialize subcontent slice metadata: {}", e))?;

        let mut meta_file = File::create(&metadata_path)
            .map_err(|e| format!("Failed to create subcontent metadata file '{}': {}", metadata_path, e))?;
        meta_file.write_all(json.as_bytes())
            .map_err(|e| format!("Failed to write subcontent metadata file '{}': {}", metadata_path, e))?;
        // File closed on drop, triggering WADUP processing

        Ok(())
    }
}

File: wadup-web/docker/go/wadup-guest-go/subcontent.go
package wadup

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"
)

var (
	subcontentMu      sync.Mutex
	subcontentCounter int
)

// subContentMetadata represents metadata for bytes emission
type subContentMetadata struct {
	Filename string `json:"filename"`
}

// subContentSliceMetadata represents metadata for slice emission
type subContentSliceMetadata struct {
	Filename string `json:"filename"`
	Offset   int64  `json:"offset"`
	Length   int64  `json:"length"`
}

// EmitBytes emits sub-content bytes for recursive processing.
//
// Writes data to /subcontent/data_N.bin and metadata to /subcontent/metadata_N.json.
// WADUP processes the sub-content when the metadata file is closed.
func EmitBytes(data []byte, filename string) error {
	subcontentMu.Lock()
	n := subcontentCounter
	subcontentCounter++
	subcontentMu.Unlock()

	dataPath := fmt.Sprintf("/subcontent/data_%d.bin", n)
	metadataPath := fmt.Sprintf("/subcontent/metadata_%d.json", n)

	// Write data file first
	dataFile, err := os.Create(dataPath)
	if err != nil {
		return fmt.Errorf("failed to create subcontent data file '%s': %w", dataPath, err)
	}
	if _, err := dataFile.Write(data); err != nil {
		dataFile.Close()
		return fmt.Errorf("failed to write subcontent data file '%s': %w", dataPath, err)
	}
	dataFile.Close()

	// Write metadata file (triggers processing when closed)
	metadata := subContentMetadata{Filename: filename}
	jsonData, err := json.Marshal(metadata)
	if err != nil {
		return fmt.Errorf("failed to serialize subcontent metadata: %w", err)
	}

	metaFile, err := os.Create(metadataPath)
	if err != nil {
		return fmt.Errorf("failed to create subcontent metadata file '%s': %w", metadataPath, err)
	}
	defer metaFile.Close()

	if _, err := metaFile.Write(jsonData); err != nil {
		return fmt.Errorf("failed to write subcontent metadata file '%s': %w", metadataPath, err)
	}

	return nil
}

// EmitSlice emits a slice of the input content as sub-content (zero-copy).
//
// The slice references a range of the original /data.bin content without copying.
// Only writes metadata to /subcontent/metadata_N.json.
func EmitSlice(offset, length int64, filename string) error {
	subcontentMu.Lock()
	n := subcontentCounter
	subcontentCounter++
	subcontentMu.Unlock()

	metadataPath := fmt.Sprintf("/subcontent/metadata_%d.json", n)

	metadata := subContentSliceMetadata{
		Filename: filename,
		Offset:   offset,
		Length:   length,
	}
	jsonData, err := json.Marshal(metadata)
	if err != nil {
		return fmt.Errorf("failed to serialize subcontent slice metadata: %w", err)
	}

	metaFile, err := os.Create(metadataPath)
	if err != nil {
		return fmt.Errorf("failed to create subcontent metadata file '%s': %w", metadataPath, err)
	}
	defer metaFile.Close()

	if _, err := metaFile.Write(jsonData); err != nil {
		return fmt.Errorf("failed to write subcontent metadata file '%s': %w", metadataPath, err)
	}

	return nil
}
